File: python/m2m_sentinel/langchain_tool.py
"""
LangChain BaseTool Integration for M2M Sentinel.
Compatible with LangChain Core, LangGraph, and Python Agent Frameworks.
"""

from typing import Type, Optional
from .client import M2MSentinelClient

try:
    from pydantic import BaseModel, Field
except ImportError:
    class BaseModel:
        pass
    def Field(*args, **kwargs):
        return None

try:
    from langchain_core.tools import BaseTool
except ImportError:
    # Graceful fallback if langchain_core is not pre-installed
    class BaseTool:
        name: str = ""
        description: str = ""
        args_schema: Optional[Type[BaseModel]] = None

        def _run(self, *args, **kwargs):
            raise NotImplementedError()


class AuditContractInput(BaseModel):
    address: str = Field(..., description="0x-prefixed 40-hex Base smart contract address to inspect.")


class M2MSentinelAuditTool(BaseTool):
    name: str = "m2m_audit_contract"
    description: str = (
        "Deterministic EVM bytecode capability analysis and proxy resolution for Base contracts. "
        "Execute before signing or sending transactions to unverified contracts on Base."
    )
    args_schema: Type[BaseModel] = AuditContractInput
    client: Optional[M2MSentinelClient] = None

    def __init__(self, api_key: Optional[str] = None, base_url: Optional[str] = None, **kwargs):
        super().__init__(**kwargs)
        self.client = M2MSentinelClient(api_key=api_key, base_url=base_url)

    def _run(self, address: str) -> str:
        try:
            res = self.client.audit_contract(address)
            return str(res)
        except Exception as e:
            return f"Error executing M2M Sentinel audit: {str(e)}"

    async def _arun(self, address: str) -> str:
        # Synchronous fallback in async context
        return self._run(address)

File: langchain_tool.js
'use strict';

/**
 * LangChain & LangGraph Tool Provider for M2M Sentinel.
 *
 * Compatible with @langchain/core/tools StructuredTool and DynamicStructuredTool.
 */

const { M2MSentinelClient } = require('./index.js');

class M2MSentinelLangChainTools {
  /**
   * @param {Object} [options]
   * @param {string} [options.apiKey] M2M Sentinel API Key
   * @param {string} [options.baseUrl] M2M Sentinel Base URL
   */
  constructor(options = {}) {
    this.client = new M2MSentinelClient(options);
  }

  /**
   * Return a list of tool definitions compatible with LangChain agents.
   * Can be passed directly to `createOpenAIToolsAgent({ tools })` or `AgentExecutor`.
   */
  getTools() {
    return [
      {
        name: 'm2m_audit_contract',
        description: 'Deterministic EVM bytecode capability analysis and proxy resolution for Base contracts. Run before signing or sending transactions to unverified contracts.',
        schema: {
          type: 'object',
          properties: {
            address: {
              type: 'string',
              description: 'The 0x-prefixed 40-hex Ethereum address on Base (Chain ID 8453) to inspect.'
            }
          },
          required: ['address']
        },
        func: async ({ address }) => {
          try {
            const res = await this.client.auditContract(address);
            return JSON.stringify(res);
          } catch (err) {
            return JSON.stringify({ error: err.message, status: err.status || 500 });
          }
        }
      },
      {
        name: 'm2m_get_gas_metrics',
        description: 'Fetch live Base network gas metrics and execution recommendations.',
        schema: {
          type: 'object',
          properties: {}
        },
        func: async () => {
          try {
            const res = await this.client.getGasFees();
            return JSON.stringify(res);
          } catch (err) {
            return JSON.stringify({ error: err.message, status: err.status || 500 });
          }
        }
      },
      {
        name: 'm2m_get_token_price',
        description: 'Fetch verified Base DEX token price for allowlisted assets (e.g. USDC, WETH).',
        schema: {
          type: 'object',
          properties: {
            symbol: {
              type: 'string',
              description: 'The asset symbol (e.g., USDC, WETH).'
            }
          },
          required: ['symbol']
        },
        func: async ({ symbol }) => {
          try {
            const res = await this.client.getTokenPrice(symbol);
            return JSON.stringify(res);
          } catch (err) {
            return JSON.stringify({ error: err.message, status: err.status || 500 });
          }
        }
      }
    ];
  }
}

module.exports = {
  M2MSentinelLangChainTools
};
